File: backend/src/main/java/com/munichweekly/backend/model/Submission.java
package com.munichweekly.backend.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.DecimalMax;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.Positive;
import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "submissions")
public class Submission {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(optional = false)
    private User user;

    @ManyToOne(optional = false)
    private Issue issue;

    private String imageUrl;

    @Column(length = 200)
    private String description;

    private Boolean isCover = false;

    // pending, approved, rejected, selected
    private String status = "pending";

    private LocalDateTime submittedAt = LocalDateTime.now();

    private LocalDateTime reviewedAt;
    
    // **NEW FIELDS: Image dimensions for masonry layout optimization**
    
    /**
     * Original image width in pixels
     * Used for masonry layout calculations to avoid frontend dimension fetching
     * Null for legacy submissions without dimension data
     */
    @Column(name = "image_width")
    @Positive(message = "Image width must be positive")
    private Integer imageWidth;
    
    /**
     * Original image height in pixels
     * Used for masonry layout calculations to avoid frontend dimension fetching
     * Null for legacy submissions without dimension data
     */
    @Column(name = "image_height")
    @Positive(message = "Image height must be positive")
    private Integer imageHeight;
    
    /**
     * Precomputed aspect ratio (width/height)
     * Used for efficient wide image detection in masonry layouts
     * Stored as decimal for precision and database indexing
     * Null for legacy submissions without dimension data
     */
    @Column(name = "aspect_ratio", precision = 10, scale = 6)
    @DecimalMin(value = "0.1", message = "Aspect ratio must be at least 0.1")
    @DecimalMax(value = "10.0", message = "Aspect ratio must be at most 10.0")
    private BigDecimal aspectRatio;

    public Submission() {}

    public Submission(User user, Issue issue, String imageUrl, String description) {
        this.user = user;
        this.issue = issue;
        this.imageUrl = imageUrl;
        this.description = description;
        this.status = "pending";
        this.submittedAt = LocalDateTime.now();
    }
    
    /**
     * Constructor with image dimensions for new submissions
     * Automatically calculates aspect ratio from width and height
     */
    public Submission(User user, Issue issue, String imageUrl, String description, 
                     Integer imageWidth, Integer imageHeight) {
        this(user, issue, imageUrl, description);
        this.setImageDimensions(imageWidth, imageHeight);
    }
    
    // **NEW METHODS: Image dimension handling**
    
    /**
     * Set image dimensions and automatically calculate aspect ratio
     * Ensures consistency between width, height, and aspect ratio
     * 
     * @param width  Image width in pixels (must be positive)
     * @param height Image height in pixels (must be positive)
     * @throws IllegalArgumentException if width or height is invalid
     */
    public void setImageDimensions(Integer width, Integer height) {
        if (width != null && height != null) {
            if (width <= 0 || height <= 0) {
                throw new IllegalArgumentException("Image dimensions must be positive");
            }
            this.imageWidth = width;
            this.imageHeight = height;
            this.aspectRatio = BigDecimal.valueOf((double) width / height)
                    .setScale(6, BigDecimal.ROUND_HALF_UP);
        } else {
            // If either dimension is null, clear all dimension data
            this.imageWidth = null;
            this.imageHeight = null;
            this.aspectRatio = null;
        }
    }
    
    /**
     * Check if this submission has complete dimension data
     * Used to determine whether to use stored dimensions or fetch dynamically
     * 
     * @return true if width, height, and aspect ratio are all available
     */
    public boolean hasDimensionData() {
        return imageWidth != null && imageHeight != null && aspectRatio != null;
    }
    
    /**
     * Check if this image is considered "wide" for masonry layout
     * Uses the standard 16:9 threshold (1.778 aspect ratio)
     * 
     * @return true if aspect ratio >= 16:9, false otherwise or if no dimension data
     */
    public boolean isWideImage() {
        if (aspectRatio == null) {
            return false;
        }
        return aspectRatio.compareTo(BigDecimal.valueOf(16.0 / 9.0)) >= 0;
    }
    
    /**
     * Get aspect ratio as double for calculations
     * 
     * @return aspect ratio as double, or null if no dimension data
     */
    public Double getAspectRatioAsDouble() {
        return aspectRatio != null ? aspectRatio.doubleValue() : null;
    }

    // Getters & setters ...

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public Issue getIssue() {
        return issue;
    }

    public void setIssue(Issue issue) {
        this.issue = issue;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public void setImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public LocalDateTime getSubmittedAt() {
        return submittedAt;
    }

    public void setSubmittedAt(LocalDateTime submittedAt) {
        this.submittedAt = submittedAt;
    }

    public LocalDateTime getReviewedAt() {
        return reviewedAt;
    }

    public void setReviewedAt(LocalDateTime reviewedAt) {
        this.reviewedAt = reviewedAt;
    }

    public Boolean getCover() {
        return isCover;
    }

    public void setCover(Boolean cover) {
        isCover = cover;
    }
    
    // **NEW GETTERS AND SETTERS: Image dimensions**
    
    public Integer getImageWidth() {
        return imageWidth;
    }
    
    public void setImageWidth(Integer imageWidth) {
        this.imageWidth = imageWidth;
        // Recalculate aspect ratio if both dimensions are available
        if (imageWidth != null && imageHeight != null) {
            this.aspectRatio = BigDecimal.valueOf((double) imageWidth / imageHeight)
                    .setScale(6, BigDecimal.ROUND_HALF_UP);
        }
    }
    
    public Integer getImageHeight() {
        return imageHeight;
    }
    
    public void setImageHeight(Integer imageHeight) {
        this.imageHeight = imageHeight;
        // Recalculate aspect ratio if both dimensions are available
        if (imageWidth != null && imageHeight != null) {
            this.aspectRatio = BigDecimal.valueOf((double) imageWidth / imageHeight)
                    .setScale(6, BigDecimal.ROUND_HALF_UP);
        }
    }
    
    public BigDecimal getAspectRatio() {
        return aspectRatio;
    }
    
    public void setAspectRatio(BigDecimal aspectRatio) {
        this.aspectRatio = aspectRatio;
    }
}

File: frontend/src/types/submission.ts
// Submission status enum
export enum SubmissionStatus {
  PENDING = "pending",
  APPROVED = "approved",
  REJECTED = "rejected",
  SELECTED = "selected"
}

// Issue type
export interface Issue {
  id: number;
  title: string;
  description: string;
  submissionStart: string;
  submissionEnd: string;
  votingStart: string;
  votingEnd: string;
  createdAt: string;
}

// Image dimensions interface for optimized masonry layout
export interface ImageDimensions {
  width: number;
  height: number;
  aspectRatio: number;
}

// Submission type
export interface Submission {
  id: number;
  imageUrl: string;
  description: string;
  status: SubmissionStatus;
  submittedAt: string;
  reviewedAt?: string;
  isCover: boolean;
  voteCount: number;
  issue: Issue;
  userId: number;
  userVote?: 'up' | 'down' | null; // User's voting status
  
  // **NEW: Image dimension fields for optimized masonry layout**
  // These fields are populated from backend for new submissions
  // and provide instant layout calculation without frontend dimension fetching
  imageWidth?: number;      // Original image width in pixels
  imageHeight?: number;     // Original image height in pixels  
  aspectRatio?: number;     // Precomputed aspect ratio (width/height)
}

// Paginated response type
export interface PaginatedResponse<T> {
  content: T[];
  pageable: {
    pageNumber: number;
    pageSize: number;
  };
  totalElements: number;
  totalPages: number;
}

// Submission list response type
export type SubmissionListResponse = PaginatedResponse<Submission>;

// Submission request type
export interface SubmissionRequest {
  issueId: number;
  description: string;
}

// My submission response type
export interface MySubmissionResponse {
  id: number;
  imageUrl: string;
  description: string;
  status: string;
  submittedAt: string;
  voteCount: number;
  selected: boolean;
  isCover: boolean;
  userEmail?: string;
  userNickname?: string;
  userId?: number;
  user?: {
    id: number;
    email: string;
    nickname: string;
    avatarUrl?: string;
  };
  nickname?: string;
  
  // **NEW: Image dimension fields for MySubmissionResponse**
  imageWidth?: number;
  imageHeight?: number;
  aspectRatio?: number;
}

// Admin submission response type
export interface AdminSubmissionResponse {
  id: number;
  imageUrl: string;
  description: string;
  status: SubmissionStatus | string;
  submittedAt: string;
  reviewedAt?: string;
  isCover: boolean;
  voteCount: number;
  userId: number;
  userEmail: string;
  userNickname: string;
  userAvatarUrl?: string;
  
  // **NEW: Image dimension fields for AdminSubmissionResponse**
  imageWidth?: number;
  imageHeight?: number;
  aspectRatio?: number;
}

// Utility function to check if submission has stored dimensions
export function hasStoredDimensions(submission: Submission | MySubmissionResponse | AdminSubmissionResponse): boolean {
  return !!(submission.imageWidth && submission.imageHeight && submission.aspectRatio);
}

// Utility function to get dimensions from submission
export function getSubmissionDimensions(submission: Submission | MySubmissionResponse | AdminSubmissionResponse): ImageDimensions | null {
  if (hasStoredDimensions(submission)) {
    return {
      width: submission.imageWidth!,
      height: submission.imageHeight!,
      aspectRatio: submission.aspectRatio!,
    };
  }
  return null;
}

// Utility function to determine if image is wide based on stored aspect ratio
export function isWideImage(submission: Submission | MySubmissionResponse | AdminSubmissionResponse, threshold: number = 16/9): boolean {
  if (submission.aspectRatio) {
    return submission.aspectRatio >= threshold;
  }
  return false; // Unknown dimensions, assume narrow for safe fallback
}